### DISCUSSION TITLE
WSGI transport does not close the iterable returned by the application

### DISCUSSION THREAD

**Discussion by trendels** [encode/httpx#1822] [Potential Issue] [answered]
The WSGI spec requires the gateway (httpx in this case) to call `close()` on the iterable returned by the application, if that method exists (from [PEP 3333](https://www.python.org/dev/peps/pep-3333/)):

> If the iterable returned by the application has a close() method, the server or gateway must call that method upon completion of the current request, whether the request was completed normally, or terminated early due to an application error during iteration or an early disconnect of the browser. 

httpx currently does not do this when wrapping a WSGI app. This can be verified with the [`validator` wrapper from `wsgiref.util.validate`](https://docs.python.org/3.6/library/wsgiref.html?highlight=file_wrapper#wsgiref.validate.validator):

```python
import httpx
from wsgiref.validate import validator

def application(environ, start_response):
    start_response("200 OK", [("content-type", "text/plain")])
    return [b"hello, world"]

application = validator(application)

with httpx.Client(app=application, base_url="http://localhost") as client:
    response = client.get("/")
```

Expected output: nothing

Actual output:
```
Iterator garbage collected without being closedException ignored in: <function IteratorWrapper.__del__ at 0x7ff285444d30>
Traceback (most recent call last):
  File "/usr/lib/python3.9/wsgiref/validate.py", line 295, in __del__
    assert_(self.closed,
  File "/usr/lib/python3.9/wsgiref/validate.py", line 128, in assert_
    raise AssertionError(*args)
AssertionError: Iterator garbage collected without being closed
```

I'm often using httpx to test WSGI apps (great feature btw!), and it would be good if `close` was called in accordance with the spec to test for example cleanup tasks, etc.

I can submit a PR if you agree this is an issue.

**Comment by lovelydinosaur** [comment, marked as answer]
What a very neatly summarised discussion - thank you! 😁

Yes that sounds valid. I think the place to deal with this will be adding a `close()` method to the `WSGIByteStream` class.
That'll get called into when the response is closed.
